design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,486 @@
+- 2 -> 2001:4860:4860::0010 (excluded)
+...
+- 18 -> 2001:4860:4860::001F (excluded)
+- 19 -> 2001:4860:4860::0020 
+...
+- 240 -> 2001:4860:4860::00FF
+
+Reseting `status.allocated-offset` field will trigger a Reconcile on the
+`NodeIPPool` and `allocated-offsets` will be re-calculated.
+In this process the reason will be "ReconcilingPool" and `Available`
+marked as unknown, so users can wait on it.
+
+```yaml
+  conditions:
+  - type: Available
+    status: Unknown
+    reason: ReconcilingPool
+    message: The pool is being updated
+```
+
+Here are a pair of possible failures from the pool so it's clear how the
+conditions work
+
+Exhausted IP Pool:
+```yaml
+  conditions:
+  - type: Available
+    status: False
+    reason: ExhaustedIPPool
+    message: ""
+    message: The IP pool has being depleted, manual intervention needed to modify NodeIPPool.
+```
+
+Failed re-creating the pool so the are not in sync with NNCE:
+```yaml
+  conditions:
+  - type: Available
+    status: False
+    reason: UnsyncronizedIPPool
+    message: Failed re constructing the IP pool, it may be not in sync with nodes.
+```
+
+The modified/created IP pool has overlaps on the range with another one.
+
+```yaml
+  conditions:
+  - types: Available
+    status: False
+    reason: OverlapingIPPool
+    message: The NodeIPPool XYZ is overlapping with this one.
+```
+
+To assign an IP address from a pool to an interface using NNCP, a special
+capture syntax is introduced understood by kubernetes-nmstate
+`ippool.[name].allocate` that will trigger the ip allocation mechanism.
+It will inject the result on the captured state passed to nmpolicy so it can be
+consumed by it, the `ippool.[name].allocate` tag will return a struct with the
+following fields `ip` and `prefix-length`.
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: static-ip
+capture:
+    address-from-traffic-1: ippool.traffic-1.allocate
+    address-from-traffic-2: ippool.traffic-2.allocate
+desiredState:
+    interfaces:
+    - name: eth1
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-1.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-1.prefix-length }}"
+        dhcp: false
+        enabled: true
+    - name: eth2
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: "{{ capture.address-from-traffic-2.ip }}"
+          prefix-length: "{{ capture.address-from-traffic-2.prefix-length }}"
+        dhcp: false
+        enabled: true
+```
+
+The allocated IPs will be marked at the `NodeNetworkConfigurationEnactment`
+status and will be used to re-construct the nmstate on reboot and to discover
+if new allocation if needed or to de-allocate them the pool is no longer
+referenced, this is done on distributed fashion by each handler at their nodes.
+
+The `NNCE.status.ippool` information allows to reconstruct `NodeIPPool.status.allocated-offsets`
```

Comment:
I still don't get it. If you change the "begin" you should update the offsets, right. But you have the offset list, you know what change was done to the begin, so you can re-calculate the offsets. Why in this case do you need to know what node it using the IP?

Reply:
In case we continue allowing `NodeIPPool` update, it's true that offset can be recalculated comparing old a new begin but since this work is going to be distribuged across handlers doing the logic of just registering the IPs they have looks simpler.